### ISSUE TITLE
Feature Request - Allow deployment to production environment during pull_request

### ISSUE THREAD

**Issue by rtrindvg** [Azure/static-web-apps#750]
**Is your feature request related to a problem? Please describe.**
I am currently facing a problem adjusting my workflow using the GitHub Actions SWA deploy component. In my company we have 2 subscriptions, one for testing and another for production. Our current workflow always works by deploying the components to the testing subscription during the pull_request event and deploys to production in another pull_request event (using labels), followed by an automated PR merge (so not on the push event). This makes it impossible to deploy it to the production environment on the testing subscription. It is also a problem during the production deployment on prod sub, since only on the push event the action is expected to update it.

**Describe the solution you'd like**
I would like to have an optional parameter where I can specify I don't want the default behavior and treat the push and pull_request events in exactly the same way. This way I can put those interactions exactly where I want them on my workflow.

**Describe alternatives you've considered**
I tried using the production subscription with the staging feature, but this creates a lot of dependencies with other components which do not have this staging feature, making everything too complex to setup without specific changes made to the software just for making it work.
I could also change my workflow to deploy to the production subscription during the push event, but this creates an issue where if something goes wrong on the deployment, the code is already merged to the main branch. This creates another set of problems for us (we prefer the philosophy of deploying first and merging only after it is successful, avoiding these conflicts).

**Comment by diberry** [comment]
Is this something that is unique to SWA or is it more an issue with your pipeline?

**Comment by rtrindvg** [comment]
Hello, @diberry. Thanks for the quick comment. This is unique to SWA, as far as I know. The rest of my environment uses AKS and I can freely create my workflow as specified (2 independent subscriptions using pull_request events) without any issues. 

The way the current SWA actions component is structured leaves my without any options to make it work on my current workflows, and we really do not want to redesign them entirely because of SWA.

**Comment by anthonychu** [comment]
@rtrindvg Can you please share a sample workflow and point out where you're having problems?

**Comment by rtrindvg** [comment]
Not sure I can make this clear with a sample workflow but I can try:
```yaml
name: Sample workflow
on:
  pull_request:
    types: [opened, synchronize, reopened, labeled]
    branches:
      - 'main'
    paths:
      - 'ui/my-frontend/**'
      - 'my-backend/**'

jobs:
  ###############
  # WORKFLOW 01 #
  ###############
  start-validations:
    if: |
      github.event.action != 'labeled'
      && github.event.pull_request.draft == false
  # Removed for brevity, linting, unittest and build unrelated to the issue

  ###############
  # WORKFLOW 02 #
  ###############
  # Removed elements from this section related to the label management and the backend deployment

  web-app-deploy-dev:
    if: |
      github.event.action == 'labeled'
      && github.event.pull_request.draft == false
      && github.event.label.name == '✨👍a02-approved-for-dev-deploy'
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          submodules: true

      - uses: Azure/static-web-apps-deploy@v1
        with:
          azure_static_web_apps_api_token: ${{ secrets.azure_static_web_apps_api_token }}
          repo_token: ${{ secrets.gh_token }}
          action: 'upload'
          app_location: 'ui/my-frontend'
          api_location: ''
          output_location: 'out'

  ###############
  # WORKFLOW 03 #
  ###############  
  web-app-deploy-prod:
    if: |
      github.event.action == 'labeled'
      && github.event.pull_request.draft == false
      && github.event.label.name =='🔥👍b02-approved-for-prod-deploy'
  # Almost the same as the workflow 02, except for an automerge at the end.
```

As you can see, my current workflow is broken in 3 sequence of actions:
- On pull_request open, sync and reopen events, I start the workflow, validating the PR without using any environment (staging or prod).
- On pull_request label events, if there is a label approving the deploy to development subscription, we want to publish it there. This is my problem number one. Since I am on a pull_request event, the action creates my deployment in a staging environment inside the app. This works, but it's not ideal, since we want to use the "production" environment of the test subscription.
- On the pull_request label event with the production approval label, we have the same issue but now on production. We REALLY do not want to use the staging feature here. We are ready to deploy it. If I change this to a push event, this leads me to an entire rabbit hole of not having an ideal way to deal with failures on the workflow and the commit is already merged to main. (eg. failure on the Cypress end-2-end testing after deployment).

Having an option on the action to allow me to deploy to the production environment on the pull_request event would work perfectly with my entire setup without any major changes.

Another path would be to have another way to upload an SWA, without GitHub integration. As far as I could research, there is no way to do this as of now. Also tried to check for the action to evaluate how hard it would be to change it, but it seems to be closed source.

Hope it is clearer now. Feel free to ask more questions is this is still unclear. I am definitely not an expert on both tools, so if I am missing a critical piece of the puzzle, any help is appreciated.

**Comment by mishapos** [comment]
@rtrindvg It's possible to achieve the behavior you want by modifying your workflow file to include this to the call to the Azure/static-web-apps-deploy@v1 task:

```yaml
env:
  DEPLOYMENT_PROVIDER: "Custom"
  REPOSITORY_URL: ${{github.event.repository.html_url}}
  BRANCH: ${{github.event.pull_request.head.ref}} 
uses: Azure/static-web-apps-deploy@v1
  with:
    azure_static_web_apps_api_token: ${{ secrets.azure_static_web_apps_api_token }}
    repo_token: ${{ secrets.gh_token }}
    # Rest omitted for brevity
```

By setting the `DEPLOYMENT_PROVIDER` environment variable to something other than `GitHub`, you are telling the action that you wish to handle setting all the deployment configuration values yourself, and you don't want it to automatically deploy to a preview environment even though the deployment was triggered from a pull request. Thus, you have to manually specify the repo URL and the branch. By default, with just a branch and a repo URL, the Azure/static-web-apps-deploy@v1 action will deploy to the production environment of the static web app matching the Api token you specified. 

This should enable you to deploy to the production environment even during a pull request. Please let me know if it works for you, and thank you for using Static Web Apps. 😄

**Comment by Areej-Ashraf** [comment]
I'm getting some sort of the same issue.
Can we pass argument type 'env production' directly deploy to my production environment for my development static web app for every PR opened. 
I want to use the production environment one as backend API is only attached to that one.
  ```
    - name: Deploy
        uses: Azure/static-web-apps-deploy@v1
        with:
          azure_static_web_apps_api_token: ${{ env.swa_key }}
          repo_token: ${{ secrets.GITHUB_TOKEN }}
          action: "upload"
          app_location: "/"
          output_location: "dist"
          arguments: "--env production"
```

**Comment by rtrindvg** [comment]
I never went back to say the suggestion from mishapos worked perfectly. I suggest Areej-Ashraf to open a new issue if this is still a problem for you. Closing...
